torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2910,6 +2911,152 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+
+  ExprPtr simplified_cmp_select_expr = nullptr;
+  if ((simplified_lhs == lhs) && (simplified_rhs == rhs) &&
+      (simplified_ret1 == ret1) && (simplified_ret2 == ret2)) {
+    simplified_cmp_select_expr = v;
+  } else {
+    simplified_cmp_select_expr = alloc<CompareSelect>(
+        simplified_lhs,
+        simplified_rhs,
+        simplified_ret1,
+        simplified_ret2,
+        v->compare_select_op(),
+        v->bias());
+  }
+
+  GRAPH_DEBUG(
+      "(SimplifierUnderContext) after simplify: ",
+      std::to_string(simplified_cmp_select_expr));
+
+  BoundInfo lhs_bound;
+  BoundInfo rhs_bound;
+  auto lhs_has_bound = getBoundInfo(simplified_lhs, lhs_bound);
+  auto rhs_has_bound = getBoundInfo(simplified_rhs, rhs_bound);
+  if (!lhs_has_bound || !rhs_has_bound) {
+    GRAPH_DEBUG(
+        "(SimplifierUnderContext) Final: ",
+        std::to_string(simplified_cmp_select_expr));
+    return simplified_cmp_select_expr;
+  }
+
+  analysis::BoundCompareResult cmp_res = analysis::BoundCompareResult::kEQ;
+  auto bound_solved = analysis::compareBound(
+      {lhs_bound.first, lhs_bound.second},
+      {rhs_bound.first, rhs_bound.second},
+      cmp_res);
+  if (!bound_solved) {
+    GRAPH_DEBUG(
+        "(SimplifierUnderContext) Final: ",
+        std::to_string(simplified_cmp_select_expr));
+    return simplified_cmp_select_expr;
+  }
+
+  auto _not = [](const CompareSelectOperation& cmp_op) {
+    switch (cmp_op) {
+      case CompareSelectOperation::kGT:
+        return CompareSelectOperation::kLE;
+      case CompareSelectOperation::kGE:
+        return CompareSelectOperation::kLT;
+      case CompareSelectOperation::kLT:
+        return CompareSelectOperation::kGE;
+      case CompareSelectOperation::kLE:
+        return CompareSelectOperation::kGT;
+      case CompareSelectOperation::kNE:
+        return CompareSelectOperation::kEQ;
+      default:
+        TORCH_INTERNAL_ASSERT(cmp_op == CompareSelectOperation::kEQ)
+        return CompareSelectOperation::kNE;
+    }
+  };
+
+  auto _imply = [](const analysis::BoundCompareResult& cmp_res,
+                   const CompareSelectOperation& cmp_op) {
+    // a > b
+    auto gt_true = (cmp_res == analysis::BoundCompareResult::kGT);
+    // a >= b
+    auto ge_true =
+        (cmp_res == analysis::BoundCompareResult::kGE ||
+         cmp_res == analysis::BoundCompareResult::kGT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a < b
+    auto lt_true = (cmp_res == analysis::BoundCompareResult::kLT);
+    // a <= b
+    auto le_true =
+        (cmp_res == analysis::BoundCompareResult::kLE ||
+         cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a == b
+    auto eq_true = (cmp_res == analysis::BoundCompareResult::kEQ);
+    // a != b
+    auto ne_true =
+        (cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kGT);
```

Comment:
This has to include `kNE` as well, right?

Reply:
The `compareBound` does not return `kNE` because it is implied by `kLT` and `kGT`. But I think it makes sense to add `kNE` here to make it easier to read.